server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -201,12 +205,36 @@ public void restoreSnapshot(final RestoreSnapshotRequest request, final ActionLi
                 // Make sure that we can restore from this snapshot
                 validateSnapshotRestorable(repositoryName, snapshotInfo);
 
+                SnapshotException snapshotException = null;
+                Metadata globalMetadata = null;
                 // Resolve the indices from the snapshot that need to be restored
-                final List<String> indicesInSnapshot = filterIndices(snapshotInfo.indices(), request.indices(), request.indicesOptions());
+                try {
+                    globalMetadata = repository.getSnapshotGlobalMetadata(snapshotId);
+                } catch (SnapshotException e) {
+                    snapshotException = e;
```

Comment:
This looks broken, we should always be able to read global metadata (we always write at least a dummy instance of it to the repository). If we fail here, then that means the snapshot is corrupted and we should throw right away and can simplify the below code accordingly.

Reply:
I followed `SharedClusterSnapshotRestoreIT.testRestoreSnapshotWithCorruptedGlobalState`. It targets situation when global metadata is broken, not empty. 
It makes sense to me that even when global metadata is broken we should be able to restore indices and data for rescue purposes. If we throw here (as it was before) we can't do that even when we don't care about data streams. WDYT?

Reply:
Ah I see ... but then we should only try to load the global metadata if either `includeGlobalState()` or data streams have been requested for restore specifically?
That way we don't create some new weird half-working state where indices are restored but data streams aren't when both are requested. Also, we save loading the global metadata needlessly this way.
I guess that needs a data streams parameter on the restore request or at least a flag data streams yes/no?

I don't think we should introduce an API here that would return a `200` even though data streams couldn't be restored. No one ever looks at the logs :)

Reply:
Ok, I took another turn on that, storing data streams' names in userMeta (similarly as SLM stores policy name there). 
It simplifies things greatly (no need to change `Entry` and its serialization etc) and saves us from loading global state on every restore (we need it only on restoring data streams).
It does enlarge `SnapshotInfo`, it should not be that bad though as they are only names not whole data streams. 
Let me know what do you think?